### DISCUSSION TITLE
Using the compiler in React Server Components (in Next.js 13)

### DISCUSSION THREAD

**Discussion by joulev** [mdx-js/mdx#2168] [Q&A] [answered]
How to make the compiler work for React Server Components? I'm using Next.js 13's new `app` directory, but I think this also applies to RSC in general.

Right now, with this (and some `webpack` configuration to import `.mdx` files as raw strings),

```mdx
This is a test **Markdown** file
```

```tsx
async function getContent() {
  const rawContent = await import(`./test.mdx`).then(x => x.default);
  return rawContent;
}

export default async function Page() {
  const content = await getContent();
  return <div>{content}</div> /* not parsing yet for inspection purposes */;
}
```

I can get the Markdown file to be displayed raw:

<img width="686" alt="image" src="https://user-images.githubusercontent.com/44609036/200168112-08e910be-3505-4b7e-a229-8ebccb19e2c5.png">

But when I add [the compiler](https://mdxjs.com/guides/mdx-on-demand/#quick-example) in,

```tsx
import { compile } from "@mdx-js/mdx";

async function getContent() {
  const rawContent = await import(`./test.mdx`).then(x => x.default);
  const content = String(await compile(rawContent, { outputFormat: "function-body" }));
  return content;
}

export default async function Page() {
  const content = await getContent();
  return <div>{content}</div> /* not parsing yet for inspection purposes */;
}
```

a very mysterious error message appears:

<img width="909" alt="image" src="https://user-images.githubusercontent.com/44609036/200168169-9606dbb4-8599-412c-8919-5f2324710efa.png">

How to resolve this error and use the compiler in React Server Components in general?

**Comment by wooorm** [comment, marked as answer]
I think this is about https://github.com/Rich-Harris/periscopic/issues/10, which was solved more than a year ago. I’m guessing you are not using the latest `@mdx-js/mdx`.

**Reply by wooorm** [reply]
Could also be that [this export map](https://github.com/Rich-Harris/periscopic/blob/2ff43ecd3e60a660acce552046d5c6eac0586ae7/package.json#L9-L11) is messing with webpack?

I don’t see the relation with RSC.

I’m also not sure why you’re using a dynamic import of an `.mdx` file, and get a string back? If you know which file to import, use a static import statement at the top of the file. If you don’t, you should probably `fs.readFile` it instead of `import()`

**Reply by joulev** [reply]
Thanks for the speedy response!

Hmm, I just checked `periscopic` in my local `node_modules`, and it is v3.0.4, which is the latest version of `periscopic` already. So any fixes from the linked issue should already be in my `node_modules`.

Also I realised I didn't have `@mdx-js/mdx` installed directly via `package.json`, but instead I had `@mdx-js/react` and `@mdx-js/loader`, both at latest version too (v2.1.5). After installing `@mdx-js/mdx` v2.1.5 manually the error still persists.

**Reply by joulev** [reply]
> I’m also not sure why you’re using a dynamic import of an `.mdx` file, and get a string back? If you know which file to import, use a static import statement at the top of the file. If you don’t, you should probably `fs.readFile` it instead of `import()`

In the actual app I'm using a dynamic path there (so `test` in `test.mdx` is not static), that's why I'm dynamically importing it. Either way since I can get the file content as string, I don't think it matters.

**Reply by wooorm** [reply]
Can you remove the entire `exports` object from `package.json` of `periscopic` locally?
Note: you might also have multiple versions in different places (`npm ls periscopic` should show them)

**Reply by wooorm** [reply]
>  I don't think it matters.

Currently not no. But, it’s confusing, because `imports` typically don’t return a string. Importing MDX should expose a component.

**Reply by joulev** [reply]
> Can you remove the entire `exports` object from `package.json` of `periscopic` locally?

No it didn't work.

**Reply by wooorm** [reply]
Then I need a small reproduction to help you.
It sounds like a bug in the particular tools that Next.js 13 introduces though. It should support `periscopic`. It’s not a bug in this project.

**Reply by joulev** [reply]
Hi, you can use this to reproduce the problem: https://github.com/joulev/debug/tree/nextjs-42534

**Reply by wooorm** [reply]
It does solve the issue. Either a) removing the `exports`, b) renaming `import` to `default`, or c) using just a string (`"exports": "./src/index.js",`).

However, the problem persists as it exists in 3 packages: `periscopic`, `estree-walker`, and `is-reference`.
You can (and likely should?) raise this issue with Next.js folks. They should likely set the `import` condition and so support `import` in export maps.
You can also PR to the three projects. I’m guessing @Rich-Harris would appreciate your help here, because Node.js’ docs say that there should always be a `default` set: https://nodejs.org/api/packages.html#conditional-exports.
In the mean time, you could perhaps use [`patch-package`](https://github.com/ds300/patch-package#readme)?

**Reply by joulev** [reply]
Thank you. Issues raised at Next.js: vercel/next.js#42534, although I'm not sure if I could explain the bug clearly enough. Since there are likely other packages that may crash too (other than the three you mentioned), I think it's best to let the Next.js team fix this once and for all.
